Full Site Editing: Add Support for Templates Default and Custom Titles and Descriptions

Copons opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #25755 

Update the templates handling to allow for custom titles and descriptions.

- Created a `theme-provided` (the name is bad, I'm open to suggestions!) custom status for `wp_template` CPT to indicate those templates that have been imported from files and not touched by the user. Previously we were using `auto-draft`, which wasn't super appropriate, and created a few complications (e.g. autodrafts can't have titles, aren't supposed to show up anywhere, etc.).
- Added support for `excerpts` to the `wp_template` CPT. Right now titles and excerpts are only editable when editing the template in its own post editor.
- Added a `theme` post meta, containing the source theme for `theme-provided` templates.
- Updated the Templates wp-admin list to display more details (title, slug, excerpt, status, source), so that now it's at least slightly useful. It also now lists `theme-provided` and (at least temporarily) `auto-draft` templates.
- Created a list of default template types definitions in PHP, containing titles and descriptions of all the "generic" (not containing variables in their slugs) templates.
- The list is filterable with the new `template_types_definitions` filter, which is already used to temporarily remove the `embed` definition, since it's currently not supported.
- The list is available in JS via Redux: it's in `settings.defaultTemplateTypesDefinitions` of `core/edit-site`.
- Templates imported from files will automatically obtain the title and excerpt from the default definitions, if possible.
- New templates created from the Site Editor sidebar will also automatically use the default definitions, and will be saved as `draft`. They will become `publish`ed once saved the first time.
- The default definitions are used across the Site Editor as a fallback for templates without title or excerpt.

### Follow up

- Update the template definitions copy!
- Add a way to edit the title (and maybe excerpt too) of a template from the Site Editor (e.g. in the Document Settings dropdown).
- Extend the wp-admin Templates list's quick edit with the excerpt (and maybe the slug too?).
- Update the template saving logic to allow for going back and forth between `draft` and `publish`.

## How has this been tested?

Note: this will create new templates which are only accessible with the changes included in this PR. Changing branch would make the new templates inaccessible; they shouldn't be used by the site, but are still there, and may cause unexpected conflicts with pre-existing `auto-draft` templates.

1. Open the Site Editor. Behind the scenes, all theme's and plugins' templates will be converted into new `wp_template` items of status `theme-provided`; this means that there will be duplicates.
2. Open the wp-admin Templates list (`/edit.php?post_type=wp_template`).
3. Make sure it shows the following columns: title, slug, description, status.
4. Notice it shows also `auto-draft` templates. Feel free to wipe this list clean by moving all templates to trash. New `theme-provided` templates will be generated again once opening the Site Editor.
5. Open the Site Editor (if you wiped the templates, now it will generate new ones, hopefully without duplicates).
6. Open the sidebar, and navigate into the Templates menu.
7. Make sure the items use the title and description as provided by the definitions list introduced in this PR.
8. Create a new template with the New Template dropdown (click on `+` in the Templates menu).
9. Make sure the items in the New Template dropdown use the definitions list.
10. Make sure the new template's name is the one provided by the definitions list (it should match the one from the New Template dropdown).
11. Check in the sidebar that the new template shows up (it might be easier to just look in the All Templates menu), it has the correct title and description, and is marked as "[Draft]".
12. Click on "Update Design" (it might be disabled: do some changes to enable it).
13. The template has now been published, so make sure it lost the "[Draft]" label.
14. Open the wp-admin Templates list, and make sure the new template shows up with the right title, description, and status.

## Screenshots

The new Templates wp-admin list:

<img width="1185" alt="Screenshot 2020-10-28 at 12 01 10" src="https://user-images.githubusercontent.com/2070010/97433370-4d7d5880-1915-11eb-9805-6bf2d6991ae8.png">

The Site Editor sidebar using the new default definitions, and including a draft template:

<img width="522" alt="Screenshot 2020-10-26 at 17 44 59" src="https://user-images.githubusercontent.com/2070010/97208552-29a20180-17b3-11eb-892b-753f84693ffe.png">

The Site Editor documents settings using the new default definitions:

<img width="276" alt="Screenshot 2020-10-26 at 17 44 41" src="https://user-images.githubusercontent.com/2070010/97208602-3888b400-17b3-11eb-9848-9e6662943acc.png">

## Types of changes
Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Copons committed:
Add handling for template excerpts

Copons committed:
Add default template types definitions

Copons committed:
Save file-based templates using the templates definitions and with theme-provided status

Copons commented on lib/load.php:
```diff
@@ -103,6 +103,7 @@ function gutenberg_is_experiment_enabled( $name ) {
 
 // Include FSE related files only if the experiment is enabled.
 if ( gutenberg_is_experiment_enabled( 'gutenberg-full-site-editing' ) ) {
+	require dirname( __FILE__ ) . '/full-site-editing/templates-definitions.php';
```
We are starting to have a lot of FSE-related library files, so I think it's the case to group them in a separate folder.
I haven't moved the others to avoid bloating this PR.

Copons committed:
Improve handling of custom post status

Copons committed:
Use the PHP templates definitions before falling back to the JS ones

mtias commented on lib/full-site-editing/templates-definitions.php:
```diff
@@ -0,0 +1,85 @@
+	return array(
+		'index'          => array(
+			'title'         => 'Default (Index)',
+			'description'   => 'Main template, applied when no other template is found',
```
Let's make them translatable from the start

Copons committed:
Localize template types definitions

Copons committed:
Make the definitions filterable

Copons committed:
Update the new template dropdown to use the new status

Copons committed:
Improve names and clean up

Copons committed:
Completely replace the JS definitions with the PHP ones

Copons committed:
New templates use definitions too

Copons committed:
Fix templates wp-admin list

Copons committed:
Sort out merge issues

Copons committed:
Replace getTemplateInfo with useTemplateInfo

Copons committed:
Prevent editor crash

Copons commented:
When I've replaced `getTemplateInfo` with the `useTemplateInfo` hook, I've also removed the former's tests.
This is because I'm not exactly sure how to test a hook, which also relies on `useSelect` to work. 🤔 
I'd be happy to restore the tests if anyone could give me a pointer on where to ~copy~ look for other hooks tests.

Copons committed:
Add useTemplateInfo tests

Addison-Stavlo reviewed (commented):
I like where this is going.

The testing steps worked as expected except:

> (if you wiped the templates, now it will generate new ones, hopefully without duplicates)

I got duplicates following the steps the first time through.  After I got through all the steps I went back and tried another round of trashing / reloading site editor and it seemed to work without duplicates. 🤷‍♀️ 

Do you think this `theme-provided` status should be used for template-parts as well? 🤔

Copons commented:
> The testing steps worked as expected except:
> 
> > (if you wiped the templates, now it will generate new ones, hopefully without duplicates)
> 
> I got duplicates following the steps the first time through. After I got through all the steps I went back and tried another round of trashing / reloading site editor and it seemed to work without duplicates. 🤷‍♀️

@Addison-Stavlo The first run can be a bit messy for many reasons.
For example, if you try to delete all templates before checking out this branch (e.g. by selecting all and moving to trash), chances are the autodrafts won't be deleted as they don't show up in the list, and so can't be selected.

The idea is to temporarily show `auto-draft` templates everywhere (site editor and wp-admin list) to have a clear idea of what is what.
If we follow through with the custom status, I'd hide `auto-draft` again, restoring its intended behaviour: automatic saves triggered by the editor, only used for revisions.

> Do you think this `theme-provided` status should be used for template-parts as well? 🤔

Yes!
But in a follow up, if we think this is the right way to go.

---

Also, we need to be 100% sure about the new status name, because it will be extraordinarily messy to change later on (see for example the duplicated templates you're seeing when testing this PR 🙂 ).
`theme-provided` was my first idea, but it's not great (e.g. plugin can provide templates as well).

Ideally the name should convey that it originates from a file (`file-based`?), leaving the actual source (theme? which theme? plugin? which plugin?) to the post meta.

Copons committed:
Add a theme source to templates

Copons committed:
Fix source for customized theme templates

Copons commented:
Similarly to `wp_template_part`, I've added a new `theme` post meta to `wp_template`.

When a template file provided by a theme is converted into a `wp_template`, it stores the theme's slug (`stylesheet`) in this new `theme` post meta.

_(Note: at the moment we only convert templates provided by the current theme, hence we simply get the current theme's stylesheet.)_

This "source" meta is displayed in the `wp_template` wp-admin list.
`theme-provided` templates' source will be "Theme: {theme name or slug}"; all other templates (edited by user) will simply show "Site".
This means that if a `theme-provided` template is edited by a user, and its status turns into `publish`, it's source will be "Site" and not "Theme" anymore".

Addison-Stavlo reviewed (commented):
> Similarly to wp_template_part, I've added a new theme post meta to wp_template.

Great!

> all other templates (edited by user) will simply show "Site".
>  This means that if a theme-provided template is edited by a user, and its status turns into publish, it's source will be "Site" and not "Theme" anymore".

I think we should try to make templates and template-parts consistent with how they behave regarding this sort of thing.

Currently, when a template-part is created by a user its theme is labeled as `custom`.  I think either we should update template parts to be 'site' as well, or change it here to be 'custom' as well.

The other difference is that this only happens for new template parts.  Template parts that originally come from a theme file will still have its theme's name saved in the meta once it goes from auto-draft to publish.  I think this is by design as it corresponds to a template part from a specific theme that you have made customizations to, but the information about the theme it came from still exists as that information may still be useful.

If we are going with a status such as`theme-provided`, then we know that `publish` state is customized even if its theme meta still references the theme in which it came from.  And maybe 'site' or 'custom' should only be given to templates that are created from scratch, and not those that are customized from a theme supplied one? 🤔 

<hr/>

> Yes!
But in a follow up, if we think this is the right way to go.

Sounds good 👍 

> Also, we need to be 100% sure about the new status name
theme-provided was my first idea, but it's not great (e.g. plugin can provide templates as well).
Ideally the name should convey that it originates from a file (file-based?),

Ah yeah, I'm not great with coming up with names either.  Maybe something as simple as `unedited` would imply it came from an outside source and hasn't been changed?

Copons commented:
> > all other templates (edited by user) will simply show "Site".
> > This means that if a theme-provided template is edited by a user, and its status turns into publish, it's source will be "Site" and not "Theme" anymore".
> 
> I think we should try to make templates and template-parts consistent with how they behave regarding this sort of thing.
> 
> Currently, when a template-part is created by a user its theme is labeled as `custom`. I think either we should update template parts to be 'site' as well, or change it here to be 'custom' as well.
> 
> The other difference is that this only happens for new template parts. Template parts that originally come from a theme file will still have its theme's name saved in the meta once it goes from auto-draft to publish. I think this is by design as it corresponds to a template part from a specific theme that you have made customizations to, but the information about the theme it came from still exists as that information may still be useful.
> 
> If we are going with a status such as`theme-provided`, then we know that `publish` state is customized even if its theme meta still references the theme in which it came from. And maybe 'site' or 'custom' should only be given to templates that are created from scratch, and not those that are customized from a theme supplied one? 🤔

I've definitely failed to explain how the meta is used. 🤦 

When a user edits a `theme-provided` template, making it their own, the `theme` meta is unaffected, and keeps its original value (which is: the theme that provided it in the first place).

Then, the Source column in the `wp_template` wp-admin list will show it as "Site" rather than "Theme: Foobar".
The code should make this "switcheroo" clear: https://github.com/WordPress/gutenberg/blob/4b9eb4b6bdfe6ec8e3f0401d2b1286a25ce2317e/lib/templates.php#L207-L217

I guess this might be improved a little bit. 🙂 
- If `theme-provided`: source is "Theme: Foobar".
- If fully custom: source is "Custom".
- If customized from a theme-provided template: source is "Custom - Originally from theme: Foobar"

> Currently, when a template-part is created by a user its theme is labeled as `custom`. I think either we should update template parts to be 'site' as well, or change it here to be 'custom' as well.

That's not a very consistent behaviour...
For example, you can create a template part from the `wp_template_part` wp-admin list, and it would show up with `theme: ''`.

IMHO the empty string should be the way to go here, rather than using `custom` (or any other arbitrary string).
An empty string automatically means that the element doesn't belong to any theme, and has the additional side effect of not conflicting with an hypothetical theme named `custom`. 😄 

Templates are already like this: fully custom templates will be `theme: ''`, while theme-provided ones (even once customized) will have the theme slug (stylesheet).

Addison-Stavlo reviewed (commented):
That makes sense, thanks for clarifying! 

>I guess this might be improved a little bit. 🙂
> * If theme-provided: source is "Theme: Foobar".
> * If fully custom: source is "Custom".
> * If customized from a theme-provided template: source is "Custom - Originally from theme: Foobar"

Yeah that amount of distinction makes sense to me 👍 .  (Maybe "Theme: Foobar (customized)" to keep it shorter if need be.)

> IMHO the empty string should be the way to go here, rather than using custom (or any other arbitrary string).

I think you have convinced me that the empty string would also make more sense in the template-part case!  This will be more consistent and we can always label them as 'Custom' in the interface if the theme meta is blank.

Copons committed:
Improve the template list columns output

Copons commented:
Replaced by #26636
